src/date_utils.ts

```diff
@@ -0,0 +1,68 @@
+export function addDaysToDate(date: Date, days: number): Date {
+    return new Date(date.getFullYear(), date.getMonth(), date.getDate() + days, date.getHours(), date.getMinutes(), date.getSeconds(), date.getMilliseconds());
+}
+
+export function addWeeksToDate(date: Date, weeks: number): Date {
+    return addDaysToDate(date, weeks * 7)
+}
+
+export function startOfDay(date: Date): Date {
+    const newDate = new Date(date)
+    newDate.setHours(0, 0, 0, 0)
+    return newDate
+}
+
+export function endOfDay(date: Date): Date {
+    const newDate = new Date(date)
+    newDate.setHours(23, 59, 59, 999)
+    return newDate
+}
+
+export function endOfMonth(date: Date): Date {
+    const newDate = new Date(date)
+    newDate.setMonth(newDate.getMonth() + 1, 0)
+    newDate.setHours(23, 59, 59, 999)
+    return newDate
+}
+
+export function endOfYear(date: Date): Date {
+    const newDate = new Date(date)
+    newDate.setMonth(11, 31)
+    newDate.setHours(23, 59, 59, 999)
+    return newDate
+}
+
+export function getDayDifference(date1: Date, date2: Date): number {
+    const diff = date1.getTime() - date2.getTime()
+    return Math.floor(diff / (1000 * 60 * 60 * 24))
+}
+
+export function getWeekDifference(date1: Date, date2: Date): number {
+    const diff = date1.getTime() - date2.getTime()
+    return Math.floor(diff / (1000 * 60 * 60 * 24 * 7))
+}
+
+/**
+ * 
+ * @returns 1 for Monday, 2 for Tuesday, etc. 7 for Sunday
+ */
+export function getIsoWeekDay(date: Date): number {
+    // 0 is Sunday, 1 is Monday, etc.
+    return date.getDay() === 0 ? 7 : date.getDay()
+}
+
+export function setIsoWeekDay(date: Date, day: number): Date {
+    const distance = day - getIsoWeekDay(date)
+    return addDaysToDate(date, distance)
+}
+
+export function getDayOfYear(date: Date): number {
+    const millisecondsInDay = 1000 * 60 * 60 * 24
+    return (Date.UTC(date.getFullYear(), date.getMonth(), date.getDate()) - Date.UTC(date.getFullYear(), 0, 0)) / millisecondsInDay;
+}
+
+export function toFormattedString(date: Date): string {
+    return new Date(date.getTime() - (date.getTimezoneOffset() * 60000))
+        .toISOString()
+        .split("T")[0];
```

Comment:
I couldn't understand what that means
```javascript
(date.getTimezoneOffset() * 60000)
```

Reply:
This function returns the date string in `YYYY-MM-DD` format. To formatting we use `toISOString()` which returns a string in `YYYY-MM-DDTHH:mm:ss.sssZ` format. Then we split by T and take first part.
However, there is a problem. `toISOString()` result is not in Date object's timezone. [Returned string's timezone is always zero UTC offset](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toISOString). That is why we need to construct a new Date in a timezone with zero UTC offset. We get current timezoneoffset by `date.getTimezoneOffset()` which is in minutes and multiply it with 60 * 1000 to convert to milliseconds.